veps/sig-storage/vm-object-graph.md

```diff
@@ -0,0 +1,258 @@
+          "labels": {
+            "type": "storage"
+          },
+          "optional": false,
+          "children": []
+        }
+      ]
+    }
+  ]
+}
+```
+
+### **User Flow**
+
+1. Access the ObjectGraph API through the subresource endpoint for a VM/VMI.
+2. Parse the response and filter unnecessary objects (e.g., in backup scenarios).
+3. Use the retrieved data as needed.
+
+### **Included Resources**
+
+- **InstanceType (`spec.instancetype.name`)**
+  - **ControllerRevision (`spec.instancetype.revisionName`)**
+- **Preference (`spec.preference.name`)**
+  - **ControllerRevision (`spec.preference.revisionName`)**
+- **VirtualMachineInstance (VMI):** Identified by VM name.
+  - **Virt-launcher Pod:** Identified by label.
+  - **Volumes**:
+    - **DataVolumes (`spec.template.spec.volumes[*].dataVolume`)**
+    - **PersistentVolumeClaims (`spec.template.spec.volumes[*].persistentVolumeClaim`)**
+    - **ConfigMaps (`spec.template.spec.volumes[*].configMap`)**
+    - **Secrets (`spec.template.spec.volumes[*].secret`)**
+    - **ServiceAccounts (`spec.template.spec.volumes[*].serviceAccount`)**
+    - **MemoryDump (`spec.template.spec.volumes[*].memoryDump`)**
+  - **AccessCredentials**
+    - **SSH Secrets (`spec.template.spec.accessCredentials.sshPublicKey.source.secret`)**
+    - **User Password Secrets (`spec.template.spec.accessCredentials.userPassword.source.secret`)**
+
+**Backend Storage PVC**  
+Identified by the persistent state PVC label.
+
+**Other Resources:**
+- Should we include `networkAttachmentDefinitions` and `networks` in the ObjectGraph?
+- Should optional objects such as `VMExports` or `VMSnapshots` be considered?
+
+## **Alternatives**
+
+## Alternatives Considered
+
+### Flat Dependency List
+
+An alternative was to return a **flat list** of dependent objects without hierarchical relationships.
+
+Example schema:
+
+```go
+type ObjectGraphNode struct {
+    ObjectReference k8sv1.TypedObjectReference `json:"objectReference"`
+    Labels          map[string]string          `json:"labels,omitempty"`
+}
+```
+
+Example output:
+
+```json
+{
+  "items": [
+    {
+      "objectReference": {
+        "apiGroup": "kubevirt.io",
+        "kind": "VirtualMachineInstance",
+        "name": "vm1",
+        "namespace": "default"
+      }
+    },
+    {
+      "objectReference": {
+        "apiGroup": "",
+        "kind": "Pod",
+        "name": "virt-launcher-vm1",
+        "namespace": "default"
+      }
+    },
+    {
+      "objectReference": {
+        "apiGroup": "cdi.kubevirt.io",
+        "kind": "DataVolume",
+        "name": "dv1",
+        "namespace": "default"
+      }
+    }
+  ]
+}
+```
+
+We chose the hierarchical representation because:
+
+- Dependency relationships between objects are meaningful and should be explicit.
+- Migration, backup, and restore processes often require processing objects in dependency order.
+- Hierarchical graphs allow easier extensibility (for example, marking optional nodes).
```

Comment:
What does this mean?

Reply:
There's an `optional` field in the hierarchical API that tells whether an object is optional for the VM to work or not. This would allow extend the graph with objects not yet included like VMSnapshots or VMExports.